File: aat-android/src/main/java/ch/bailu/aat/views/OsmApiEditorView.java
package ch.bailu.aat.views;

import android.content.Context;
import android.text.TextUtils;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;

import java.util.Locale;

import ch.bailu.aat_lib.search.poi.OsmApiConfiguration;
import ch.bailu.aat.util.ui.AppTheme;
import ch.bailu.aat.util.ui.UiTheme;
import ch.bailu.aat.views.description.mview.MultiView;
import ch.bailu.aat.views.preferences.TitleView;
import ch.bailu.aat.views.preferences.VerticalScrollView;

public class OsmApiEditorView extends LinearLayout {
    private final EditTextTool editor;
    private final TextView preview;

    private final MultiView inputMultiView;


    public OsmApiEditorView(Context context, OsmApiConfiguration osmApi, UiTheme theme) {
        super(context);
        setOrientation(LinearLayout.VERTICAL);

        preview = new TextView(getContext());

        VerticalScrollView scroller = new VerticalScrollView(getContext());
        scroller.add(preview);
        editor = new EditTextTool(new TagEditor(getContext(),
                osmApi.getBaseDirectory()), LinearLayout.VERTICAL, theme);

        inputMultiView = new MultiView(getContext(), osmApi.getApiName());
        inputMultiView.add(editor);
        inputMultiView.add(scroller);

        preview.setOnClickListener(view -> inputMultiView.setNext());
        preview.setText(osmApi.getUrlPreview(editor.edit.getText().toString()));

        theme.content(preview);
        addView(createTitle(osmApi, theme));
        addView(inputMultiView);

    }


    private View createTitle(OsmApiConfiguration osmApi, UiTheme theme) {
        LinearLayout layout = new LinearLayout(getContext());
        layout.setOrientation(LinearLayout.HORIZONTAL);

        String[] strings = osmApi.getUrlStart().split(osmApi.getApiName().toLowerCase(Locale.ROOT));

        TextView b = new TitleView(getContext(), osmApi.getApiName(), AppTheme.search);
        b.setSingleLine();

        if (strings.length>1) {
            TextView a = new TextView(getContext());
            TextView c = new TextView(getContext());

            a.setText(strings[0]);
            a.setSingleLine();

            c.setText(strings[1]);
            c.setSingleLine();
            c.setEllipsize(TextUtils.TruncateAt.END);

            theme.content(a);
            theme.content(c);
            b.setTextColor(theme.getHighlightColor());

            layout.addView(a);
            layout.addView(b);
            layout.addView(c);
        } else {
            layout.addView(b);
        }

        layout.setOnClickListener(view -> {
            preview.setText(osmApi.getUrlPreview(editor.edit.getText().toString()));
            inputMultiView.setNext();
        });

        return layout;
    }

    @Override
    @NonNull
    public String toString() {
        return editor.edit.getText().toString();
    }

    public void insertLine(String s) {
        editor.insertLine(s);
        inputMultiView.setActive(0);
    }

    public void setText(String query) {
        editor.edit.setText(query);
        inputMultiView.setActive(0);
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/views/osm/OsmApiEditorView.kt
package ch.bailu.aat.views.osm

import android.content.Context
import android.text.TextUtils
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import ch.bailu.aat.util.ui.theme.AppTheme
import ch.bailu.aat.util.ui.theme.UiTheme
import ch.bailu.aat.views.description.mview.MultiView
import ch.bailu.aat.views.preferences.TitleView
import ch.bailu.aat.views.preferences.VerticalScrollView
import ch.bailu.aat_lib.coordinates.BoundingBoxE6
import ch.bailu.aat_lib.api.ApiConfiguration

class OsmApiEditorView(context: Context, private val bounding: BoundingBoxE6, osmApi: ApiConfiguration, theme: UiTheme) :
    LinearLayout(context) {
    private val editor: EditTextTool
    private val preview: TextView
    private val inputMultiView: MultiView

    init {
        orientation = VERTICAL
        preview = TextView(getContext())
        val scroller = VerticalScrollView(getContext())
        scroller.add(preview)
        editor = EditTextTool(
            TagEditor(
                getContext(),
                osmApi.baseDirectory
            ), VERTICAL, theme
        )
        inputMultiView = MultiView(getContext(), osmApi.apiName)
        inputMultiView.add(editor)
        inputMultiView.add(scroller)
        preview.setOnClickListener { inputMultiView.setNext() }
        preview.text = osmApi.getUrlPreview(editor.edit.text.toString(), bounding)
        theme.content(preview)
        addView(createTitle(osmApi, theme))
        addView(inputMultiView)
    }

    private fun createTitle(osmApi: ApiConfiguration, theme: UiTheme): View {
        val layout = LinearLayout(context)
        layout.orientation = HORIZONTAL
        val strings = osmApi.urlStart.split(osmApi.apiName.lowercase().toRegex())
            .dropLastWhile { it.isEmpty() }
            .toTypedArray()
        val b: TextView = TitleView(context, osmApi.apiName, AppTheme.search)
        b.setSingleLine()
        if (strings.size > 1) {
            val a = TextView(context)
            val c = TextView(context)
            a.text = strings[0]
            a.setSingleLine()
            c.text = strings[1]
            c.setSingleLine()
            c.ellipsize = TextUtils.TruncateAt.END
            theme.content(a)
            theme.content(c)
            b.setTextColor(theme.getHighlightColor())
            layout.addView(a)
            layout.addView(b)
            layout.addView(c)
        } else {
            layout.addView(b)
        }
        layout.setOnClickListener {
            preview.text = osmApi.getUrlPreview(editor.edit.text.toString(), bounding)
            inputMultiView.setNext()
        }
        return layout
    }

    override fun toString(): String {
        return editor.edit.text.toString()
    }

    fun insertLine(text: String) {
        editor.insertLine(text)
        inputMultiView.setActive(0)
    }

    fun setText(query: String) {
        editor.edit.setText(query)
        inputMultiView.setActive(0)
    }
}
